pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go

```diff
@@ -78,155 +76,212 @@ func (r *AtlasDatabaseUserReconciler) Reconcile(context context.Context, req ctr
+
 		return result.ReconcileResult(), nil
 	}
-	ctx.Client = atlasClient
+	workflowCtx.Client = atlasClient
+
+	owner, err := customresource.IsOwner(databaseUser, r.ObjectDeletionProtection, managedByOperator(), managedByAtlas(ctx, atlasClient, project.ID(), log))
+	if err != nil {
+		result = workflow.Terminate(workflow.Internal, fmt.Sprintf("enable to resolve ownership for deletion protection: %s", err))
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+		log.Error(result.GetMessage())
 
-	result = r.ensureDatabaseUser(ctx, *project, *databaseUser)
-	if !result.IsOk() {
-		ctx.SetConditionFromResult(status.DatabaseUserReadyType, result)
 		return result.ReconcileResult(), nil
 	}
 
-	ctx.SetConditionTrue(status.DatabaseUserReadyType)
-	ctx.SetConditionTrue(status.ReadyType)
-	return result.ReconcileResult(), nil
-}
+	if !owner {
+		result = workflow.Terminate(
+			workflow.AtlasDeletionProtection,
+			"unable to reconcile database user: it already exists in Atlas, it was not previously managed by the operator, and the deletion protection is enabled.",
+		)
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+		log.Error(result.GetMessage())
 
-func (r *AtlasDatabaseUserReconciler) readProjectResource(user *mdbv1.AtlasDatabaseUser, project *mdbv1.AtlasProject) workflow.Result {
-	if err := r.Client.Get(context.Background(), user.AtlasProjectObjectKey(), project); err != nil {
-		return workflow.Terminate(workflow.Internal, err.Error())
+		return result.ReconcileResult(), nil
 	}
-	return workflow.OK()
-}
 
-func (r *AtlasDatabaseUserReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	c, err := controller.New("AtlasDatabaseUser", mgr, controller.Options{Reconciler: r})
-	if err != nil {
-		return err
+	deletionRequest, result := r.handleDeletion(ctx, databaseUser, project, atlasClient, log)
+	if deletionRequest {
+		return result.ReconcileResult(), nil
 	}
 
-	// Watch for changes to primary resource AtlasDatabaseUser & handle delete separately
-	err = c.Watch(&source.Kind{Type: &mdbv1.AtlasDatabaseUser{}}, &watch.EventHandlerWithDelete{Controller: r}, r.GlobalPredicates...)
+	err = customresource.ApplyLastConfigApplied(ctx, databaseUser, r.Client)
 	if err != nil {
-		return err
+		result = workflow.Terminate(workflow.Internal, err.Error())
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+		log.Error(result.GetMessage())
+
+		return result.ReconcileResult(), nil
 	}
 
-	// Watch for DatabaseUser password Secrets
-	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, watch.NewSecretHandler(r.WatchedResources))
-	if err != nil {
-		return err
+	result = r.ensureDatabaseUser(workflowCtx, *project, *databaseUser)
+	if !result.IsOk() {
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+
+		return result.ReconcileResult(), nil
 	}
 
-	return nil
-}
+	err = customresource.ManageFinalizer(ctx, r.Client, databaseUser, customresource.SetFinalizer)
```

Comment:
non-blocking: consider consistent naming: `EnsureFinalizer`

Reply:
As it's not a controller function but actually a helper, I think there's no inconsistency with naming